Allow snapshot tap changes (#4731)
firecracker-microvm/firecracker: docs/snapshotting/network-for-clones.md

Address for up to arp cache timeout seconds. After said timeout period,
connectivity will work both ways even without an explicit flush.

### Renaming host device names

In some environments where the jailer is not being used, restoring a snapshot
may be tricky because the tap device on the host will not be the same as the tap
device that the original VM was mapped to when it was snapshotted, as when the
tap device come from a pool of such devices.

In this case you can use the network_overrides parameter to snapshot restore to
specify which network device (based on the name inside the VM, such as "eth0")
maps to which host tap device (e.g. "vmtap01").

This may require reconfiguration of the networking inside the VM so that it is
still routable externally.

> 

Reviewer: Could you please add an example here of how this can be used.

Author: Added some sample code and more detail.

Reviewer: It would be great if the example include the steps to start a microVM with a certain network config (it could be the one from "Getting started") then take a snapshot and, finally, load this snapshot in a different host with a different TAP device (show the snapshot load command with the override).

Author: The problem with this in principle is that it does require that the VM reconfigure its networking to match the expectations of the external routing. This complexity is easily managed in the real world where you will have some channel of communication from host to guest (vsocks or swapping block devices or mmds or vmgenid) but hard to capture in a small example.

Reviewer: That is fine. I think we can still add something useful here. The guest configuration you have already is fine. I would add something like that:

```
For example, if we have a network interface named `eth0` in the snapshotted microVM. We can override it during snapshot resume, like this:

curl --unix-socket /tmp/firecracker.socket -i \
    -X PUT 'http://localhost/snapshot/load' \
    -H  'Accept: application/json' \
    -H  'Content-Type: application/json' \
    -d '{
            "snapshot_path": "./snapshot_file",
            "mem_backend": {
                "backend_path": "./mem_file",
                "backend_type": "File"
            },
            "enable_diff_snapshots": true,
            "resume_vm": false,
            "network_overrides": [
                 {
                     iface_id: "eth0",
                     host_dev_name": "vmtap01"
                 }
            ]
    }'
```

after this bit:

> In this case you can use the `network_overrides` parameter to snapshot restore
to specify which network device (based on the name inside the VM, such as
"eth0") maps to which host tap device (e.g. "vmtap01").

Author: Done

Address for up to arp cache timeout seconds. After said timeout period,
connectivity will work both ways even without an explicit flush.

> ### Renaming host device names

Reviewer 2: The markdown linter [seems to complain](https://buildkite.com/firecracker/firecracker-pr/builds/12273#01945105-f732-4a16-812e-07e0fdf0e5d2/345). You can fix it locally with `./tools/devtool fmt` and check with `./tools/devtool checkstyle`.

Author: Fixed

connectivity will work both ways even without an explicit flush.

### Renaming host device names

In some environments where the jailer is not being used, restoring a snapshot
may be tricky because the tap device on the host will not be the same as the tap
device that the original VM was mapped to when it was snapshotted, for example
when the tap device comes from a pool of such devices.

In this case you can use the `network_overrides` parameter of the snapshot
restore API to specify which guest network device maps to which host tap device.

For example, if we have a network interface named `eth0` in the snapshotted
microVM, we can override it to point to the host device `vmtap01` during
snapshot resume, like this:

```bash
curl --unix-socket /tmp/firecracker.socket -i \
    -X PUT 'http://localhost/snapshot/load' \
    -H  'Accept: application/json' \
    -H  'Content-Type: application/json' \
    -d '{
            "snapshot_path": "./snapshot_file",
            "mem_backend": {
                "backend_path": "./mem_file",
                "backend_type": "File"
            },
            "network_overrides": [
                 {

>                      iface_id: "eth0",
>                      host_dev_name": "vmtap01"

Reviewer 3: Sorry, missed this earlier, but this needs to be
```suggestion
                     "iface_id": "eth0",
                     "host_dev_name": "vmtap01"
```

Reviewer: :facepalm:

Reviewer: fixed

Address for up to arp cache timeout seconds. After said timeout period,
connectivity will work both ways even without an explicit flush.

### Renaming host device names

> In some environments where the jailer is not being used, restoring a snapshot
> may be tricky because the tap device on the host will not be the same as the tap
> device that the original VM was mapped to when it was snapshotted, as when the
> tap device come from a pool of such devices.

Reviewer 3: I don't quite understand the last part of this sentence. Should it be

```suggestion
In some environments where the jailer is not being used, restoring a snapshot
may be tricky because the tap device on the host will not be the same as the tap
device that the original VM was mapped to when it was snapshotted, for example when the
tap device comes from a pool of such devices.
```

?

Reviewer: I think so. I've amended it with your suggested changes.

> In this case you can use the `network_overrides` parameter to snapshot restore

Reviewer 3: ```suggestion
In this case you can use the `network_overrides` parameter during snapshot restore
```

of

```suggestion
In this case you can use the `network_overrides` parameter of the snapshot restore API
```

Reviewer: done

to specify which guest network device maps to which host tap device.

> For example, if we have a network interface named `eth0` in the snapshotted
> microVM. We can override it to point to the host device `vmtap01` during
> snapshot resume, like this:

Reviewer 3: ```suggestion
For example, if we have a network interface named `eth0` in the snapshotted
microVM, we can override it to point to the host device `vmtap01` during
snapshot resume like this:
```

```bash
curl --unix-socket /tmp/firecracker.socket -i \
    -X PUT 'http://localhost/snapshot/load' \
    -H  'Accept: application/json' \
    -H  'Content-Type: application/json' \
    -d '{
            "snapshot_path": "./snapshot_file",
            "mem_backend": {
                "backend_path": "./mem_file",
                "backend_type": "File"
            },
            "enable_diff_snapshots": true,

>             "resume_vm": false,

Reviewer 3: nit: maybe this example should focus only on the mandatory arguments + `network_overrides`. E.g. drop the `enable_diff_snapshots` and `resume_vm` and just leave them at Firecracker's defaults, because they're not important for this example

Reviewer: done

            "network_overrides": [
                 {
                     iface_id: "eth0",
                     host_dev_name": "vmtap01"
                 }
            ]
    }'
```

This may require reconfiguration of the networking inside the VM so that it is
still routable externally. The

> [network setup documentation](../network-setup.md) in the "In The Guest" section

Reviewer 3: nit: could be a direct link to the subsection

Reviewer: done

Address for up to arp cache timeout seconds. After said timeout period,
connectivity will work both ways even without an explicit flush.

### Renaming host device names

In some environments where the jailer is not being used, restoring a snapshot
may be tricky because the tap device on the host will not be the same as the tap
device that the original VM was mapped to when it was snapshotted, as when the
tap device come from a pool of such devices.

> In this case you can use the network_overrides parameter to snapshot restore to

Reviewer 2: nit: quote it `network_overrides`
